crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! It's goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which will be used by this allocator is a single page of memory, which in Wasm is
+//! 64KiB. We do not expect contracts to use more memory than this (for now), so we will throw an
+//! OOM error instead of requesting more memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is 64KiB
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+pub struct BumpAllocator;
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        INNER.alloc(layout)
+    }
+
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    ///
+    /// If the heap hasn't been initialized yet this value will be `None`.
+    next: Option<usize>,
+
+    /// The address of the upper limit of our heap.
+    ///
+    /// If the heap hasn't been initialized yet this value will be `None`.
+    upper_limit: Option<usize>,
+}
+
+impl InnerAlloc {
+    pub const fn new() -> Self {
+        Self {
+            next: None,
+            upper_limit: None,
+        }
+    }
+
+    unsafe fn alloc(&mut self, layout: Layout) -> *mut u8 {
+        // TODO: Figure out how to properly initalize the heap
+        let alloc_start = if let Some(start) = self.next {
+            start
+        } else {
+            let prev_page = core::arch::wasm32::memory_grow(0, 1);
+            if prev_page == usize::max_value() {
+                panic!("OOM")
```

Comment:
You won't be able to see this message anywhere I am afraid. We are compiling the core lib so that it just dies silently on panic in order to prevent all the fmt stuff to be pulled it. Additionally, it is recommended to not `panic` here but rather return a nullptr or call `handle_alloc_error`:  https://doc.rust-lang.org/stable/core/alloc/trait.GlobalAlloc.html#errors

What I would recommend is the following: Call `ink_env::debug_message` with your error string and then return a null pointer. This way you can have the error messages when enabling `ink_env/ink-debug`.

The same es true for all the places you call `expect`.

Reply:
I want to add further that I consider `expect` to be a bad use for proper error information forwarding since `expect` can only forward a `&str` and nothing more. Instead what I usually do is to call `.unwrap_or_else(|error| panic!("something bad happened: {}", error))`.

Reply:
> You won't be able to see this message anywhere I am afraid. We are compiling the core lib so that it just dies silently on panic in order to prevent all the fmt stuff to be pulled it.

Gotcha, is that because of [`cargo-contract`'s use of `panic_immediate_abort`](https://github.com/paritytech/cargo-contract/blob/50d27922d6947cf7f8630dfe0ee7b9bb5b4de249/src/cmd/build.rs#L189)?

> Additionally, it is recommended to not `panic` here but rather return a nullptr or call `handle_alloc_error`: https://doc.rust-lang.org/stable/core/alloc/trait.GlobalAlloc.html#errors

Ah, I didn't read that in the docs, thanks!

Reply:
> What I would recommend is the following: Call ink_env::debug_message with your error string and then return a null pointer. This way you can have the error messages when enabling ink_env/ink-debug.

We can't use `ink_env` in the `ink_allocator` crate or else we end up with cyclic dependencies. I think we could move the debugging code out of `ink_env`, but that's probably better left for another PR

Reply:
> Gotcha, is that because of [`cargo-contract`'s use of `panic_immediate_abort`](https://github.com/paritytech/cargo-contract/blob/50d27922d6947cf7f8630dfe0ee7b9bb5b4de249/src/cmd/build.rs#L189)?

Yes and also because the panic messages are not really forwarded from the execution engine to the client console. You will see this messages when executing in inks off-chain environment, though.